### PR TITLE
fs: merge upload_fobj and upload

### PR DESCRIPTION
**Pull request by skshetry**
* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

File: dvc/fs/base.py [hunk 1]
```diff
@@ -217,44 +217,49 @@ def is_dir_hash(cls, hash_):
...
+                "read",
+                disable=no_progress_bar,
+                bytes=True,
+                total=total,
+                desc=desc,
+                **pbar_args,
+            ) as wrapped:
+                # `size` is used to provide hints to the WebdavFileSystem
+                # for legacy servers.
+                return self.upload_fobj(wrapped, to_info, size=total)
```
Comment (skshetry) [inline dvc/fs/base.py:250]:
```suggestion
                # pylint: disable=no-member
                return self.upload_fobj(wrapped, to_info, size=total)
```

**Review by pmrowla — APPROVED** [review]

**Comment by efiop** [comment]
@skshetry Btw, could you clarify how you see it in relation to fsspec?

File: dvc/fs/base.py [hunk 2]
```diff
@@ -217,44 +217,50 @@ def is_dir_hash(cls, hash_):
...
 
         if to_info.scheme != self.scheme:
             raise NotImplementedError
 
-        if from_info.scheme != "local":
-            raise NotImplementedError
-
         logger.debug("Uploading '%s' to '%s'", from_info, to_info)
+        if is_file_obj:
+            with Tqdm.wrapattr(
```
Comment (karajan1001) [inline dvc/fs/base.py:239]: This TQDM wrapper in different levels for `upload_fobj` and `_upload`. Hard to maintain them.

Comment (skshetry) [reply dvc/fs/base.py:239]: This is temporary.

**Comment by skshetry** [comment]
I don't think I have any motivations to unify them from fsspec perspective. My first preference there would be to have an explicit API like `upload_fileobj`/`upload_fobj` and my second choice would be to extend `pipe_file` to support file objects. 
`upload` and `upload_fobj` are quite different things in the filesystem layer.

For us, what we need is a `fs.utils.write` like method (discussed in https://github.com/iterative/dvc/pull/6558#discussion_r704927903) that can also provide callback support, which is more of a service layer for fs.

I am assuming `upload_fobj` as a fsspec's API at least for now, so I have tried to not add callback support there, as the caller can do so themselves quite easily. 

I'll consider `upload` API to be that `fs.utils.write`, that extends fsspec API for us, rather than being fsspec-related.

**Pull request merged**